File: scripts/python/feature_scaling.py
#!/usr/bin/env python3

import argparse

import pandas as pd

from sklearn.preprocessing import (

    StandardScaler,

    MinMaxScaler,

    RobustScaler

)

########################################################

parser = argparse.ArgumentParser()

parser.add_argument("--train", required=True)

parser.add_argument("--test", required=True)

parser.add_argument("--method", required=True)

parser.add_argument("--metadata", required=True)

parser.add_argument("--train_output", required=True)

parser.add_argument("--test_output", required=True)

args = parser.parse_args()

########################################################

train = pd.read_csv(

    args.train,

    sep="\t"

)

test = pd.read_csv(

    args.test,

    sep="\t"

)

########################################################

meta_cols = [

    x.strip()

    for x in args.metadata.split(",")

]

meta_cols = [

    x

    for x in meta_cols

    if x in train.columns

]

if args.metadata and not meta_cols:
    raise ValueError(
        "None of the specified metadata columns were found in the training file."
    )

########################################################

feature_cols = [

    c

    for c in train.columns

    if c not in meta_cols

]

if not feature_cols:
    raise ValueError(
        "No feature columns found for scaling."
    )

# بررسی عددی بودن featureها
non_numeric = train[feature_cols].select_dtypes(
    exclude="number"
).columns.tolist()

if non_numeric:
    raise ValueError(
        f"The following feature columns are not numeric: {non_numeric}"
    )

########################################################


########################################################

if args.method == "standard":

    scaler = StandardScaler()

elif args.method == "minmax":

    scaler = MinMaxScaler()

elif args.method == "robust":

    scaler = RobustScaler()

else:

    raise ValueError("Unknown scaling method.")

########################################################

# بررسی وجود ستون‌های متادیتا در test
missing_meta = set(meta_cols) - set(test.columns)

if missing_meta:
    raise ValueError(
        f"The following metadata columns are missing in test file: {sorted(missing_meta)}"
    )

# بررسی وجود ستون‌های Feature در test
missing_features = set(feature_cols) - set(test.columns)

if missing_features:
    raise ValueError(
        f"The following feature columns are missing in test file: {sorted(missing_features)}"
    )

# بررسی عددی بودن Featureهای test
non_numeric_test = test[feature_cols].select_dtypes(
    exclude="number"
).columns.tolist()

if non_numeric_test:
    raise ValueError(
        f"The following feature columns in test are not numeric: {non_numeric_test}"
    )

# بررسی وجود مقادیر گمشده در train
if train[feature_cols].isnull().any().any():
    raise ValueError(
        "Training data contains missing values (NaN)."
    )

# بررسی وجود مقادیر گمشده در test
if test[feature_cols].isnull().any().any():
    raise ValueError(
        "Test data contains missing values (NaN)."
    )


train_scaled = scaler.fit_transform(
    train[feature_cols]
)

test_scaled = scaler.transform(
    test[feature_cols]
)




########################################################

train_scaled = pd.DataFrame(

    train_scaled,

    columns=feature_cols

)

########################################################

test_scaled = pd.DataFrame(

    test_scaled,

    columns=feature_cols

)

########################################################

train_out = pd.concat(

    [

        train[meta_cols].reset_index(drop=True),

        train_scaled

    ],

    axis=1

)

########################################################

test_out = pd.concat(

    [

        test[meta_cols].reset_index(drop=True),

        test_scaled

    ],

    axis=1

)

########################################################

train_out.to_csv(

    args.train_output,

    sep="\t",

    index=False

)

########################################################

test_out.to_csv(

    args.test_output,

    sep="\t",

    index=False

)

########################################################

print("Scaling method:", args.method)

print("Features:", len(feature_cols))

File: scripts/R/feature_scaling.R
library(caret)

########################################################
# Check arguments
########################################################

args <- commandArgs(trailingOnly = TRUE)

if (length(args) != 6) {
    stop(
        paste(
            "Usage:",
            "Rscript script.R",
            "train_file",
            "test_file",
            "method",
            "meta_cols",
            "train_out",
            "test_out"
        )
    )
}

train_file <- args[1]
test_file  <- args[2]

method <- trimws(
    strsplit(args[3], ",")[[1]]
)

meta_cols <- trimws(
    strsplit(args[4], ",")[[1]]
)

train_out <- args[5]
test_out  <- args[6]

########################################################
# Read data
########################################################

train <- read.delim(
    train_file,
    check.names = FALSE
)

test <- read.delim(
    test_file,
    check.names = FALSE
)

########################################################
# Save original column order
########################################################

train_order <- colnames(train)
test_order  <- colnames(test)

########################################################
# Check metadata columns
########################################################

missing_meta_train <- setdiff(meta_cols, colnames(train))

if (length(missing_meta_train) > 0) {
    stop(
        paste(
            "Missing metadata columns in train:",
            paste(missing_meta_train, collapse = ", ")
        )
    )
}

missing_meta_test <- setdiff(meta_cols, colnames(test))

if (length(missing_meta_test) > 0) {
    stop(
        paste(
            "Missing metadata columns in test:",
            paste(missing_meta_test, collapse = ", ")
        )
    )
}

########################################################
# Feature columns
########################################################

feature_cols <- setdiff(colnames(train), meta_cols)

if (length(feature_cols) == 0) {
    stop("No feature columns found.")
}

########################################################
# Check feature columns in test
########################################################

missing_features <- setdiff(feature_cols, colnames(test))

if (length(missing_features) > 0) {
    stop(
        paste(
            "Missing feature columns in test:",
            paste(missing_features, collapse = ", ")
        )
    )
}


########################################################
# Check extra feature columns in test
########################################################

extra_features <- setdiff(
    setdiff(colnames(test), meta_cols),
    feature_cols
)

if (length(extra_features) > 0) {
    stop(
        paste(
            "Extra feature columns in test:",
            paste(extra_features, collapse = ", ")
        )
    )
}

########################################################
# Check numeric features
########################################################

non_numeric <- feature_cols[
    !sapply(train[, feature_cols, drop = FALSE], is.numeric)
]

if (length(non_numeric) > 0) {
    stop(
        paste(
            "Non-numeric feature columns:",
            paste(non_numeric, collapse = ", ")
        )
    )
}



########################################################
# Check numeric features in test
########################################################

non_numeric_test <- feature_cols[
    !sapply(test[, feature_cols, drop = FALSE], is.numeric)
]

if (length(non_numeric_test) > 0) {
    stop(
        paste(
            "Non-numeric feature columns in test:",
            paste(non_numeric_test, collapse = ", ")
        )
    )
}

########################################################
# Warn about missing values
########################################################

if (anyNA(train[, feature_cols, drop = FALSE])) {
    warning("Training data contains missing values.")
}

if (anyNA(test[, feature_cols, drop = FALSE])) {
    warning("Test data contains missing values.")
}

########################################################
# Learn preprocessing from train
########################################################

pre <- preProcess(
    train[, feature_cols, drop = FALSE],
    method = method
)

########################################################
# Apply preprocessing
########################################################

train_scaled <- predict(
    pre,
    train[, feature_cols, drop = FALSE]
)

test_scaled <- predict(
    pre,
    test[, feature_cols, drop = FALSE]
)

########################################################
# Combine metadata and processed features
########################################################

train <- cbind(
    train[, meta_cols, drop = FALSE],
    train_scaled
)

test <- cbind(
    test[, meta_cols, drop = FALSE],
    test_scaled
)

########################################################
# Restore original column order
########################################################

train <- train[, train_order]

test <- test[, test_order]

########################################################
# Write output
########################################################

write.table(
    train,
    train_out,
    sep = "\t",
    quote = FALSE,
    row.names = FALSE
)

write.table(
    test,
    test_out,
    sep = "\t",
    quote = FALSE,
    row.names = FALSE
)